File: api/internal/models/validation.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type (
	ValidationOrigin string

	ValidationStatus string

	Validation struct {
		ID     uuid.UUID  `json:"id"`
		UserID *uuid.UUID `json:"userId,omitempty"`

		// Single validation fields
		GuestIP                  *string       `json:"-"`
		GuestUserAgent           *string       `json:"-"`
		SingleTargetEmail        *string       `json:"singleTargetEmail,omitempty"`
		SingleTargetReachability *Reachability `json:"singleTargetReachability,omitempty"`

		// Bulk validation fields
		UploadFilename   *string              `json:"uploadFilename,omitempty"`
		BulkAddressCount *int                 `json:"bulkAddressCount,omitempty"`
		ReportToken      *uuid.UUID           `json:"reportToken,omitempty"`
		ProviderSource   *ContactProviderType `json:"providerSource,omitempty"`

		Origin    ValidationOrigin `json:"origin"`
		Status    ValidationStatus `json:"status"`
		CreatedAt time.Time        `json:"createdAt"`
	}
)

const (
	AppValidation ValidationOrigin = "app"
	APIValidation ValidationOrigin = "api"

	StatusProcessing ValidationStatus = "processing"
	StatusFailed     ValidationStatus = "failed"
	StatusCompleted  ValidationStatus = "completed"
)

File: frontend/app/types/validation.ts
import { ContactProviderType } from "./contactProvider";
import { Reachability } from "./email";

export interface Validation {
    id: string,
    singleTargetEmail?: string,
    singleTargetReachability?: Reachability,
    bulkAddressCount?: number,
    uploadFilename?: string,
    reportToken?: string,
    providerSource?: ContactProviderType,
    origin: string,
    status: ValidationStatus,
    createdAt: string
}

export enum ValidationStatus {
    Completed = "completed",
    Failed = "failed",
    Processing = "processing"
}

export enum ValidationOrigin {
    Platform = "app",
    API = "api"
}